### PR TITLE
Add JSON null support for the built-in (ReadOnly)Memory converters.

### LINKED ISSUE #95267: JsonConverter for {ReadOnly}Memory doesn't deserialize nulls
Given that this code runs successfully without exception or assert:
```C#
using System.Diagnostics;
using System.Text.Json;

ReadOnlyMemory<float> rom;

rom = default;
Debug.Assert(rom.IsEmpty);

rom = null;
Debug.Assert(rom.IsEmpty);

rom = (ReadOnlyMemory<float>)null;
Debug.Assert(rom.IsEmpty);

rom = ((float[])null!).AsMemory();
Debug.Assert(rom.IsEmpty);

rom = JsonSerializer.Deserialize<float[]>("null");
Debug.Assert(rom.IsEmpty);
```
it breaks my expectations that this:
```C#
rom = JsonSerializer.Deserialize<ReadOnlyMemory<float>>("null"); // boom
```
throws an exception:
```
Unhandled exception. System.Text.Json.JsonException: The JSON value could not be converted to System.ReadOnlyMemory`1[System.Single]. Path: $ | LineNumber: 0 | BytePositionInLine: 4.
   at System.Text.Json.ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(Type propertyType)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.Deserialize(Utf8JsonReader& reader, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo`1 jsonTypeInfo, Nullable`1 actualByteCount)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo`1 jsonTypeInfo)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
```

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fix #95267.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/MemoryConverter.cs [hunk 1]
```diff
@@ -2,12 +2,30 @@
...
     {
         internal override bool CanHaveMetadata => false;
+        public override bool HandleNull => true;
+
+        internal override bool OnTryRead(
+            ref Utf8JsonReader reader,
+            Type typeToConvert,
+            JsonSerializerOptions options,
+            scoped ref ReadStack state,
+            [MaybeNullWhen(false)] out Memory<T> value)
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/MemoryConverter.cs:19]: This attribute isn't needed, is it?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ReadOnlyMemoryConverter.cs [hunk 2]
```diff
@@ -2,12 +2,30 @@
...
     {
         internal override bool CanHaveMetadata => false;
+        public override bool HandleNull => true;
+
+        internal override bool OnTryRead(
+            ref Utf8JsonReader reader,
+            Type typeToConvert,
+            JsonSerializerOptions options,
+            scoped ref ReadStack state,
+            [MaybeNullWhen(false)] out ReadOnlyMemory<T> value)
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ReadOnlyMemoryConverter.cs:19]: Ditto

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/MemoryConverter.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/MemoryConverter.cs]: Ah yes, I blame the tooling :-)

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonMetadataServicesConverter.cs [hunk 3]
```diff
@@ -29,8 +30,16 @@ internal sealed class JsonMetadataServicesConverter<T> : JsonResumableConverter<
 
         public JsonMetadataServicesConverter(JsonConverter<T> converter)
         {
-            ConverterStrategy = converter.ConverterStrategy;
             Converter = converter;
+            ConverterStrategy = converter.ConverterStrategy;
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonMetadataServicesConverter.cs:34]: I was about to ask :) Does that mean there's another bug that was lurking and this was fixing?

[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonMetadataServicesConverter.cs:34]: Correct, it was surfaced with the newly added tests.

[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonMetadataServicesConverter.cs:34]: For context this is a converter wrapper type that is being used by the source generator APIs. The logic I just added was copied from a similar wrapper type being used internally:

https://github.com/dotnet/runtime/blob/acf7c85d2ab9c598ff71b6baa733a7098f7f6148/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/CastingConverter.cs#L27-L36

**Review by tarekgh — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ReadOnlyMemoryConverter.cs [hunk 4]
```diff
@@ -2,12 +2,30 @@
...
+using System.Diagnostics.CodeAnalysis;
 
 namespace System.Text.Json.Serialization.Converters
 {
     internal sealed class ReadOnlyMemoryConverter<T> : JsonCollectionConverter<ReadOnlyMemory<T>, T>
     {
         internal override bool CanHaveMetadata => false;
+        public override bool HandleNull => true;
+
+        internal override bool OnTryRead(
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ReadOnlyMemoryConverter.cs:14]: Why do you need to override both?

**Review by jozkee — APPROVED** [review]

**Comment by Jozkee** [comment]
Should this be backported to 8.0.x?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ReadOnlyMemoryConverter.cs (hunk 4 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ReadOnlyMemoryConverter.cs:14]: `HandleNull` controls whether the serialization infrastructure should call this converter on `JsonTokenType.Null` values (by default it does that handling automatically), so we need to override it in order to run the custom logic in the `OnTryRead` method.

**Pull request merged**